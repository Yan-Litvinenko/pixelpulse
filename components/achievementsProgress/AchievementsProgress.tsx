import React from 'react';
import AchievementsChallengeBtn from '../achievementsChallengeBtn/AchievementsChallengeBtn';
import AchievementsProgressRing from '../achievementsProgressRing/AchievementsProgressRing';
import styles from '@/styles/components/achievementsProgress/AchievementsProgress.module.scss';
import { achievementsSelector } from '@/redux/selectors';
import { useSelector } from 'react-redux';

export default function AchievementsProgress(): React.JSX.Element {
    const { error, loading, amountAchieved, amountAchievements, percent } = useSelector(achievementsSelector);

    return (
        <aside className={styles.progress}>
            <div className={styles.progress__inner}>
                {loading || error ? (
                    <>
                        <AchievementsProgressRing percent={100} />
                        <span className={styles.progress__statistic}>0/0</span>
                    </>
                ) : (
                    <>
                        <AchievementsProgressRing percent={Number(percent)} />
                        <span className={styles.progress__statistic}>
                            {Number(amountAchieved)}/{Number(amountAchievements)}
                        </span>
                    </>
                )}

                <span className={styles.progress__name}>progress</span>
            </div>

            <p className={styles.progress__text}>
                I have created a set of achievements for myself and I use this page to track them.
            </p>
            <p className={styles.progress__text}>
                If you want to give me a challenge and rate it, please feel free to submit it with the button below!
            </p>

            <AchievementsChallengeBtn />
        </aside>
    );
}
