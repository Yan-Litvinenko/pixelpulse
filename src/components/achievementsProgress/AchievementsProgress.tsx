import React from 'react';
import styles from './AchievementsProgress.module.scss';
import { AchievementsProgressRing } from '../achievementsProgressRing/AchievementsProgressRing';
import { achievementsSelector } from '../../store/selectors/selectors';
import { modalOpenHandler } from '../../store/slices/modalSlice';
import { useDispatch, useSelector } from 'react-redux';
import type { AppDispatch } from '../../store/store';

const AchievementsProgress = (): React.JSX.Element => {
    const { error, loading, amountAchieved, amountAchievements, percent } = useSelector(achievementsSelector);
    const dispatch = useDispatch<AppDispatch>();

    return (
        <section className={styles.progress}>
            <div className={styles.progress__inner}>
                {loading || error ? (
                    <>
                        <AchievementsProgressRing percent={100} />
                        <span className={styles.progress__statistic}>0/0</span>
                    </>
                ) : (
                    <>
                        <AchievementsProgressRing percent={Number(percent)} />
                        <span className={styles.progress__statistic}>
                            {Number(amountAchieved)}/{Number(amountAchievements)}
                        </span>
                    </>
                )}

                <span className={styles.progress__name}>progress</span>
            </div>

            <p className={styles.progress__text}>
                I have created a set of achievements for myself and I use this page to track them.
            </p>
            <p className={styles.progress__text}>
                If you want to give me a challenge and rate it, please feel free to submit it with the button below!
            </p>

            <button
                className={styles.progress__button}
                onClick={() => dispatch(modalOpenHandler({ key: 'challenge' }))}
                type="button"
            >
                Challenge me
            </button>
        </section>
    );
};

export { AchievementsProgress };
