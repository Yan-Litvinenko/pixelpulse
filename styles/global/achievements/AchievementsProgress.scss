@import '@/styles/mixin.scss';

.achievements_progress {
    display: flex;
    flex-direction: column;
    gap: 1.5rem 0rem;
    grid-column: 1 / 2;
    grid-row: 2 / 3;
    justify-self: end;
    max-width: 18rem;

    &__inner {
        display: flex;
        flex-direction: column;
        border: 0.1rem solid var(--frame-border-color);
        height: 18rem;
        padding: 2rem 0rem;
        position: relative;
        width: 18rem;
    }

    &__ring {
        position: relative;
    }

    &__statistic {
        @include text-properties('Big-Shoulders', 2.2rem, 700);
        color: var(--main-color);
        left: 50%;
        position: absolute;
        top: 45%;
        transform: translate(-50%, -50%);
    }

    &__name {
        @include text-properties('Iceberg', 1.2rem, 400);
        bottom: 1.6rem;
        color: var(--dark-grey);
        left: 50%;
        position: absolute;
        transform: translateX(-50%);
    }

    &__text {
        @include text-properties($line-height: 1.35rem);
        color: var(--dark-grey);
    }
}

@media (max-width: 768px) {
    .achievements_progress {
        flex-direction: row;
        grid-column: 1 / -1;
        grid-row: 3 / 4;
        justify-self: start;
        max-width: 100%;
        width: 100%;

        &__inner {
            display: none;
        }

        &__text {
            display: none;
        }

        &__name {
            display: none;
        }
    }
}
