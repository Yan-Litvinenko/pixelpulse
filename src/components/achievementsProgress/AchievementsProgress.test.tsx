import React from 'react';
import { AchievementsProgress } from './AchievementsProgress';
import { render, screen, fireEvent } from '@testing-library/react';
import * as reduxHooks from 'react-redux';

jest.mock('react-redux');
const mockSelector = jest.spyOn(reduxHooks, 'useSelector');
const mockDispatch = jest.spyOn(reduxHooks, 'useDispatch');

const amountAchieved: number = 12;
const amountAchievements: number = 20;

describe('AchievementsProgress component', (): void => {
    test('Render', (): void => {
        mockSelector.mockReturnValue({ amountAchieved, amountAchievements });

        const component = render(<AchievementsProgress />);
        const statistic = screen.getByText(`${amountAchieved}/${amountAchievements}`);

        expect(statistic).toBeInTheDocument();
        expect(component).toMatchSnapshot();
    });

    test('Click challenge me', (): void => {
        const dispatch = jest.fn();
        mockDispatch.mockReturnValue(dispatch);

        render(<AchievementsProgress />);

        const button = screen.getByText(/challenge me/i);
        fireEvent.click(button);

        expect(dispatch).toHaveBeenCalledTimes(1);
    });
});
