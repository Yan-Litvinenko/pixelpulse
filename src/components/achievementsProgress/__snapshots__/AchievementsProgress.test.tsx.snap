// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AchievementsProgress component Render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section>
        <div>
          <svg
            data-testid="svg"
            height="148"
            viewBox="0 0 148 148"
            width="148"
          >
            <circle
              cx="74"
              cy="74"
              data-testid="circle"
              fill="transparent"
              r="46"
              stroke="current-color"
              stroke-width="14"
              style="stroke-dashoffset: NaN; stroke-dasharray: 289.02652413026095 289.02652413026095;"
            />
          </svg>
          <span>
            12
            /
            20
          </span>
          <span>
            progress
          </span>
        </div>
        <p>
          I have created a set of achievements for myself and I use this page to track them.
        </p>
        <p>
          If you want to give me a challenge and rate it, please feel free to submit it with the button below!
        </p>
        <button
          type="button"
        >
          Challenge me
        </button>
      </section>
    </div>
  </body>,
  "container": <div>
    <section>
      <div>
        <svg
          data-testid="svg"
          height="148"
          viewBox="0 0 148 148"
          width="148"
        >
          <circle
            cx="74"
            cy="74"
            data-testid="circle"
            fill="transparent"
            r="46"
            stroke="current-color"
            stroke-width="14"
            style="stroke-dashoffset: NaN; stroke-dasharray: 289.02652413026095 289.02652413026095;"
          />
        </svg>
        <span>
          12
          /
          20
        </span>
        <span>
          progress
        </span>
      </div>
      <p>
        I have created a set of achievements for myself and I use this page to track them.
      </p>
      <p>
        If you want to give me a challenge and rate it, please feel free to submit it with the button below!
      </p>
      <button
        type="button"
      >
        Challenge me
      </button>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
